libwallet/src/slatepack/types.rs

```diff
@@ -352,19 +441,197 @@ pub mod slatepack_version {
+	#[serde(default = "default_recipients_empty")]
+	#[serde(skip_serializing_if = "recipients_empty")]
+	recipients: Vec<SlatepackAddress>,
+}
+
+fn recipients_empty(value: &Vec<SlatepackAddress>) -> bool {
+	value.is_empty()
+}
+
+fn default_recipients_empty() -> Vec<SlatepackAddress> {
+	vec![]
+}
+
+impl SlatepackEncMetadata {
+	// return length in bytes for encoding (without the 4 byte length header)
+	pub fn encoded_len(&self) -> Result<usize, Error> {
+		let mut length = 2; //opt flags
+		if let Some(s) = &self.sender {
+			length += s.encoded_len()?;
+		}
+		if !self.recipients.is_empty() {
+			length += 1;
+			for r in self.recipients.iter() {
+				length += r.encoded_len()?;
+			}
+		}
+		Ok(length)
+	}
+}
+
+/// Wrapper for outputting encrypted metadata as binary
+#[derive(Debug, Clone)]
+pub struct SlatepackEncMetadataBin(pub SlatepackEncMetadata);
+
+impl serde::Serialize for SlatepackEncMetadataBin {
+	fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+	where
+		S: serde::Serializer,
+	{
+		let mut vec = vec![];
+		ser::serialize(&mut vec, ser::ProtocolVersion(4), self)
+			.map_err(|err| serde::ser::Error::custom(err.to_string()))?;
+		serializer.serialize_bytes(&vec)
+	}
+}
+
+impl<'de> serde::Deserialize<'de> for SlatepackEncMetadataBin {
+	fn deserialize<D>(deserializer: D) -> Result<SlatepackEncMetadataBin, D::Error>
+	where
+		D: serde::Deserializer<'de>,
+	{
+		struct SlatepackEncMetadataBinVisitor;
+
+		impl<'de> serde::de::Visitor<'de> for SlatepackEncMetadataBinVisitor {
+			type Value = SlatepackEncMetadataBin;
+
+			fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+				write!(formatter, "a serialised binary Slatepack Metadata")
+			}
+
+			fn visit_bytes<E>(self, value: &[u8]) -> Result<SlatepackEncMetadataBin, E>
+			where
+				E: serde::de::Error,
+			{
+				let mut reader = std::io::Cursor::new(value.to_vec());
+				let s = ser::deserialize(&mut reader, ser::ProtocolVersion(4))
+					.map_err(|err| serde::de::Error::custom(err.to_string()))?;
+				Ok(s)
+			}
+		}
+
+		deserializer.deserialize_bytes(SlatepackEncMetadataBinVisitor)
+	}
+}
+
+impl Writeable for SlatepackEncMetadataBin {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ser::Error> {
+		let inner = &self.0;
+		// write entire metadata length
+		writer.write_u32(inner.encoded_len().map_err(|e| {
+			error!("Cannot write encrypted metadata length: {}", e);
+			ser::Error::CorruptedData
+		})? as u32)?;
+
+		// 16 bits of optional content flags (2), most reserved for future use
+		let mut opt_flags: u16 = 0;
+		if inner.sender.is_some() {
+			opt_flags |= 0x01;
+		}
+		if !inner.recipients.is_empty() {
+			opt_flags |= 0x02;
+		}
+		writer.write_u16(opt_flags)?;
+
+		if let Some(s) = &inner.sender {
+			s.write(writer)?;
+		};
+
+		// Recipients List
+		if !inner.recipients.is_empty() {
+			let len = inner.recipients.len();
+			// write number of recipients
+			if len > 255 {
```

Comment:
No, it won't be easy to change to a u16 later.. though 255 recipients sounds excessive perhaps it might be worth a byte to ensure this is future proof